src/libraries/System.Net.WebSockets/src/System/Net/WebSockets/Compression/ZLibStreamPool.cs

```diff
@@ -0,0 +1,283 @@
+            {
+                return stream;
+            }
+
+            return CreateInflater();
+        }
+
+        public void ReturnInflater(ZLibStreamHandle stream)
+        {
+            if (stream.InflateReset() != ErrorCode.Ok)
+            {
+                stream.Dispose();
+                return;
+            }
+
+            Return(stream, _inflaters);
+        }
+
+        public ZLibStreamHandle GetDeflater()
+        {
+            if (TryGet(_deflaters, out ZLibStreamHandle? stream))
+            {
+                return stream;
+            }
+
+            return CreateDeflater();
+        }
+
+        public void ReturnDeflater(ZLibStreamHandle stream)
+        {
+            if (stream.DeflateReset() != ErrorCode.Ok)
+            {
+                stream.Dispose();
+                return;
+            }
+
+            Return(stream, _deflaters);
+        }
+
+        private void Return(ZLibStreamHandle stream, List<CacheItem> cache)
+        {
+            lock (cache)
+            {
+                if (cache.Count < MaximumRetained)
+                {
+                    cache.Add(new CacheItem(stream));
+
+                    if (Interlocked.Increment(ref _activeCount) == 1)
+                    {
+                        _cleaningTimer.Change(CacheTimeout, Timeout.InfiniteTimeSpan);
+                    }
+                    return;
+                }
+            }
+
+            // If we've reached the maximum retained capacity, we will destroy the stream.
+            // It is important that we do this outside of the cache lock.
+            stream.Dispose();
+        }
+
+        private bool TryGet(List<CacheItem> cache, [NotNullWhen(true)] out ZLibStreamHandle? stream)
+        {
+            lock (cache)
+            {
+                int count = cache.Count;
+
+                if (count > 0)
+                {
+                    CacheItem item = cache[count - 1];
+                    cache.RemoveAt(count - 1);
+                    Interlocked.Decrement(ref _activeCount);
+
+                    stream = item.Stream;
+                    return true;
+                }
+            }
+
+            stream = null;
+            return false;
+        }
+
+        private void RemoveStaleItems()
+        {
+            RemoveStaleItems(_inflaters);
+            RemoveStaleItems(_deflaters);
+
+            // There is a race condition here, were _activeCount could be decremented
+            // by a rent operation, but it's not big deal to schedule a timer tick that
+            // would eventually do nothing.
+            if (_activeCount > 0)
+            {
+                _cleaningTimer.Change(CacheTimeout, Timeout.InfiniteTimeSpan);
+            }
+        }
+
+        private void RemoveStaleItems(List<CacheItem> cache)
+        {
+            long currentTimestamp = Stopwatch.GetTimestamp();
+
+            lock (cache)
+            {
+                for (int index = cache.Count; index >= 0;)
```

Comment:
The oldest entries will always be at the front of the list right? If so, this won't be the optimization you're intending unless you happen to be clearing the entire cache. It would be better to loop from the front and then `cache.RemoveRange(0, lengthToRemove)`. For a larger cache, a ring buffer would be appropriate.

Also for a larger cache, consider what impacts there would be to 99th percentile latency -- it may be better to dispose the `Stream` outside of the lock, or to limit the number of items removed in a single pass.